docs/layout-of-source-files.rst

```diff
@@ -216,101 +216,562 @@ the code below creates new global symbols ``alias`` and ``symbol2`` which refere
+.. index:: import key, import path, filesystem path
 
-.. note::
-    Always use relative imports like ``import "./filename.sol";`` and avoid
-    using ``..`` in path specifiers. In the latter case, it is probably better to use
-    global paths and set up remappings as explained below.
+Virtual Filesystem
+~~~~~~~~~~~~~~~~~~
+
+The compiler maintains an internal database (virtual filesystem) where each compiled module is
+assigned a unique *import key* which is an opaque and unstructured identifier.
+
+While an import key in the virtual filesystem can be anything, it should be a valid path if the
+source is meant to be loaded from the underlying filesystem.
+When the requested key is not present in the virtual filesystem, it is passed to the file loader.
+In case of the command-line compiler the file loader simply uses it as a path.
+The `JavaScript interface <https://github.com/ethereum/solc-js>`_ is a bit more flexible in that
+regard and allows the user to provide a callback to perform this operation.
+In this case the keys can be arbitrary.
+For example they could be URLs as long as the custom loader can handle them.
+
+There are several ways to load modules into the virtual filesystem:
+
+#. **import statement**
+
+   The ``import`` statement tells the compiler to load and compile a module and then make its
+   content available to the current module.
+
+   We will refer to the path used in the statement as *import path*.
+   The import path is translated into an import key first and then the compiler uses the key to
+   look up the file in its virtual filesystem.
+
+   Imports can be broadly classified into three categories based on how the path is specified:
+
+   .. code-block:: solidity
+
+       import "/contracts/lib/token.sol";   // Absolute
+       import "contracts/lib/token.sol";    // Relative to base
+       import "./contracts/lib/token.sol";  // Relative to source
+
+   There is actually no distinction between absolute imports and imports relative to base at the
+   virtual filesystem level - in both cases the import path is translated into an import key using
+   the same rules - but they are handled differently by the default file loader.
+
+   Imports relative to source, on the other hand, need to be interpreted as paths and normalized to
+   be properly resolved into import keys.
+   The path in this case must conform to UNIX path conventions regardless of the underlying platform.
+
+#. **CLI**
+
+   To compile a file using the command-line interface of the compiler you specify one or more paths:
+
+   .. code-block:: bash
+
+       solc contract.sol /usr/local/dapp-bin/token.sol
+
+   These are interpreted as *filesystem paths* and the rules for translating them into import keys
+   are different than for import paths.
+   They follow the shell rules instead: they can only be relatve to
+   the current working directory and they are normalized and converted into canonical form according
```

Comment:
Maybe we should give a hint about what this means? It's mostly about converting relative into absolute paths, right?

Reply:
Done.

It's a bit more. It's also about switching the format from platform-dependent (e.g. `C:\project\contract.sol`) to the UNIX one, removing duplicate slashes and collapsing the `./` and `../` segments in the middle of the path.